### ISSUE TITLE
Create unit tests

### ISSUE THREAD

**Issue by jrgnicho** [swri-robotics/ros_scxml#18]
The following set of unit tests would help ensure that the core SM functions work as intended:
- Verify that all states in the scxml file are known to the library
- Verify that entry, exit and pre-condition callbacks can be attached to each state
- Verify Custom future class returns the expected result produced by the state entry callback
- Verify transitions 
- Verify that history state (shallow and deep) return the state machine to the corresponding state

**Comment by alexchua0108** [comment]
only manage to do up a simple test case and spend some time to understand the package

Facing issue on compiling the package:

Knowing that Debian does not package the scxml module, installation of qt513 is done and setup CMAKE_MODULE_PATH and LD_LIBRARY_PATH environment variables to
> export CMAKE_MODULE_PATH=/opt/qt513/lib/cmake:$CMAKE_MODULE_PATH
> export LD_LIBRARY_PATH=/opt/qt513/lib:$LD_LIBRARY_PATH
> export LD_LIBRARY_PATH=/opt/qt513/plugins:$LD_LIBRARY_PATH

CMake Error still remains:
> CMake Error at /usr/lib/x86_64-linux-gnu/cmake/Qt5/Qt5Config.cmake:28 (find_package):
> Could not find a package configuration file provided by "Qt5Scxml" with any
>  of the following names:

>    Qt5ScxmlConfig.cmake
>    qt5scxml-config.cmake

>  Add the installation prefix of "Qt5Scxml" to CMAKE_PREFIX_PATH or set
> "Qt5Scxml_DIR" to a directory containing one of the above files.  If
> "Qt5Scxml" provides a separate development package or SDK, be sure it has
>  been installed.
> Call Stack (most recent call first):
> CMakeLists.txt:17 (find_package)

**Comment by gavanderhoorn** [comment]
Thanks @alexchua0108 for your effort already :+1:

**Comment by dhled** [comment]
WRID2020 - I'm working on this issue

**Comment by gavanderhoorn** [comment]
@ipa-led: please note the comments in [the readme](https://github.com/swri-robotics/ros_scxml/blob/6fc916f8a35522475a8e59296ff2cf4fe4e6bda8/README.md) about the required versions of Qt.

**Comment by dhled** [comment]
> only manage to do up a simple test case and spend some time to understand the package
> 
> Facing issue on compiling the package:
> 
> Knowing that Debian does not package the scxml module, installation of qt513 is done and setup CMAKE_MODULE_PATH and LD_LIBRARY_PATH environment variables to
> 
> > export CMAKE_MODULE_PATH=/opt/qt513/lib/cmake:$CMAKE_MODULE_PATH
> > export LD_LIBRARY_PATH=/opt/qt513/lib:$LD_LIBRARY_PATH
> > export LD_LIBRARY_PATH=/opt/qt513/plugins:$LD_LIBRARY_PATH
> 
> CMake Error still remains:
> 
> > CMake Error at /usr/lib/x86_64-linux-gnu/cmake/Qt5/Qt5Config.cmake:28 (find_package):
> > Could not find a package configuration file provided by "Qt5Scxml" with any
> > of the following names:
> 
> > Qt5ScxmlConfig.cmake
> > qt5scxml-config.cmake
> 
> > Add the installation prefix of "Qt5Scxml" to CMAKE_PREFIX_PATH or set
> > "Qt5Scxml_DIR" to a directory containing one of the above files.  If
> > "Qt5Scxml" provides a separate development package or SDK, be sure it has
> > been installed.
> > Call Stack (most recent call first):
> > CMakeLists.txt:17 (find_package)

@alexchua0108 

I think there is indeed a small error in the readme. It's not CMAKE_MODULE_PATH, but CMAKE_PREFIX_PATH. I managed to build with that, I will open a PR for that. I'm actually not sure of the difference here.

**Comment by alexander-jh** [comment]
Been fighting with the same CMake error through the CRS package.

``--- stderr: crs_application                                        
CMake Error at /usr/lib/x86_64-linux-gnu/cmake/Qt5/Qt5Config.cmake:28 (find_package):
  Could not find a package configuration file provided by "Qt5Scxml" with any
  of the following names:

    Qt5ScxmlConfig.cmake
    qt5scxml-config.cmake

  Add the installation prefix of "Qt5Scxml" to CMAKE_PREFIX_PATH or set
  "Qt5Scxml_DIR" to a directory containing one of the above files.  If
  "Qt5Scxml" provides a separate development package or SDK, be sure it has
  been installed.
``
Tried sourcing like mentioned in the ros_scxml readme without any success. I've additionally tried updating the default path in qtchooser and setting Qt5Scxml's path in the CMakeLists for scxml_core.

**Comment by jrgnicho** [comment]
@alexander-jh what do you get when you echo LD_LIBRARY_PATH and CMAKE_PREFIX_PATH?
In my build I get
```
echo $LD_LIBRARY_PATH 
.../opt/ros/melodic/lib:/opt/qt513/plugins:/opt/qt513/lib
```

**Comment by alexander-jh** [comment]
@jrgnicho For both they're pointing to the right directory, but experience the same issue. For extra measure I tried unsetting, then manually setting LD_LIBRARY_PATH to be only those two locations, which didn't work either.

`echo $LD_LIBRARY_PATH   
/opt/ros/eloquent/lib:/opt/qt513/plugins:/opt/qt513/lib:`
`echo $CMAKE_PREFIX_PATH    
/opt/qt513/lib/cmake:`

**Comment by jrgnicho** [comment]
Hmm that is puzzling, I imagine when you've checked the /opt/ directory to see if qt512 it was properly installed right?

**Comment by alexander-jh** [comment]
Yes everything seems to installed from the PPA correctly.

EDIT: Finally figured out the issue. For whatever reason in `/opt/qt513/bin/qt513-env.sh` there was a conditional statement that was exporting `/opt/qt513/lib/x86_64-linux-gnu/` in the `LD_LIBRARY_PATH`. Just removing that path from the export fixed all my Qt5Scxml issues.

**Comment by Drojas251** [comment]
@alexander-jh I am having the same issue. Can you clarify your steps in solving this issue. what exactly did you remove from qt513-env.sh ? Should that line you mentioned be replaced with something else? 

Would removing that path you mentioned, then running 

export CMAKE_PREFIX_PATH=/opt/qt513/lib/cmake:$CMAKE_PREFIX_PATH
export LD_LIBRARY_PATH=/opt/qt513/lib:$LD_LIBRARY_PATH
export LD_LIBRARY_PATH=/opt/qt513/plugins:$LD_LIBRARY_PATH

fix the issue?

Any help would be greatly appreciated

**Comment by marip8** [comment]
Closing; addressed in #30
